sandbox/lua/encoders/schema_influx_write.lua

```diff
@@ -0,0 +1,319 @@
+
+--]=]
+
+require "cjson"
+require "math"
+require "string"
+
+local name_prefix_orig  = read_config("name_prefix") or "name"
+local name_prefix = name_prefix_orig
+local use_subs
+if string.find(name_prefix, "%%{[%w%p]-}") then
+    use_subs = true
+end
+
+-- Default this option to "mydb"
+local influxdb_db = read_config("database") or "mydb"
+
+-- Default this option to ""
+local retention_policy = read_config("retention_policy") or ""
+
+-- Default this option to "ms"
+local timestamp_precision = read_config("timestamp_precision") or "ms"
+
+local base_fields_map = {
+    Type = true,
+    Payload = true,
+    Hostname = true,
+    Pid = true,
+    Logger = true,
+    Severity = true,
+    EnvVersion = true
+}
+
+local base_fields_list = {
+    "Type",
+    "Payload",
+    "Hostname",
+    "Pid",
+    "Logger",
+    "Severity",
+    "EnvVersion"
+}
+
+local base_fields_tag_map = {
+    Type = true,
+    Hostname = true,
+    Severity = true,
+    Logger = true
+}
+
+-- Used for interpolating message fields into name_prefix.
+local function sub_func(key)
+    if base_fields_map[key] then
+        return read_message(key)
+    else
+        local val = read_message("Fields["..key.."]")
+        if val then
+            return val
+        end
+        return "%{"..key.."}"
+    end
+end
+
+-- Remove blacklisted fields from the set of base fields that we use, and
+-- create a table of dynamic fields to skip.
+local used_base_fields = {}
+local skip_fields_str = read_config("skip_fields")
+local skip_fields = {}
+if skip_fields_str then
+    for field in string.gmatch(skip_fields_str, "[%S]+") do
+        skip_fields[field] = true
+    end
+
+    for _, base_field in ipairs(base_fields_list) do
+        if skip_fields["**all_base**"] then
+            skip_fields[base_field] = true
+        elseif not skip_fields[base_field] then
+            used_base_fields[#used_base_fields+1] = base_field
+        else
+            skip_fields[base_field] = nil
+        end
+    end
+else
+    used_base_fields = base_fields_list
+end
+
+-- Create and populate a table of fields to be used as tags
+local tag_fields_str = read_config("tag_fields") or "**all_base**"
+local used_tag_fields = {}
+if tag_fields_str then
+    for field in string.gmatch(tag_fields_str, "[%S]+") do
+        used_tag_fields[field] = true
+    end
+end
+
+function process_message()
+    local columns = {}
+    local values = {}
+
+    -- Determine the precision of the timestamp and do the math
+    local timestamp_divisor = 1
+    if timestamp_precision == "u" then
+        timestamp_divisor = 1e3
+    elseif timestamp_precision == "ms" then
+        timestamp_divisor = 1e6
+    elseif timestamp_precision == "s" then
+        timestamp_divisor = 1e9
+    elseif timestamp_precision == "m" then
+        timestamp_divisor = 1e9 * 60
+    elseif timestamp_precision == "h" then
+        timestamp_divisor = 1e9 * 60 * 60
+    end
```

Comment:
`timestamp_divisor` value doesn't change for the lifetime of the plugin, should be calculated outside of `process_message` rather than repeated for each message.